### PR TITLE
Weblate 0.1.0

### PR DESCRIPTION
**Pull request by zhquan**
Backend to extract changes from a Weblate instance.
The changes can be collected for a given project or for all
the projects in the instance.

The initial tests have been done on:
- https://hosted.weblate.org (Weblate API 4.2)
- https://translations.documentfoundation.org (Weblate API 4.1)

Notes:
- Sleep for rate is working in >=4.1.
- Incremental fetching works for all projects in >=4.1.
- Incremental fetching works for single projects in >=4.2.
- Users details can be fetched only if the token has enough permissions.
  See details at: https://docs.weblate.org/en/latest/api.html#get--api-users-.

The backend can be executed as follows:
```
perceval weblate
https://translations.documentfoundation.org
-t
<token>
--from-date
2020-01-01
--no-archive
```

The token can be obtained after registering to a weblate
instance (e.g., https://translations.documentfoundation.org/),
via the page <instance>/accounts/profile/#api

Test added correctly.

### CODE REVIEW

File: perceval/backends/core/weblate.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+#     Valerio Cosentino <valcos@bitergia.com>
+#
+
+import logging
+
+import requests
+
+from grimoirelab_toolkit.datetime import str_to_datetime
+from grimoirelab_toolkit.uris import urijoin
+from ...utils import DEFAULT_DATETIME
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:28]: Add this line at the end of the next block of imports.

File: perceval/backends/core/weblate.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+                        DEFAULT_SEARCH_FIELD)
+from ...client import HttpClient, RateLimitHandler
+
+# Range before sleeping until rate limit reset
+MIN_RATE_LIMIT = 10
+# Default sleep time and retries to deal with connection/server problems
+DEFAULT_SLEEP_TIME = 1
+MAX_RETRIES = 5
+
+CATEGORY_CHANGES = "changes"
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:42]: Shouldn't it be better to call it `change`? That will mean we need to change the name of the variable also.

File: perceval/backends/core/weblate.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class Weblate(Backend):
+    """Weblate backend for Perceval.
+
+    This class retrieves the changes for a given list of
+    keywords via the Google API. To initialize this class a list
+    of keywords is needed.
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:53]: This text is wrong and references to Google API and not to Weblate.

File: perceval/backends/core/weblate.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+
+        kwargs = {
+            'from_date': from_date
+        }
+        items = super().fetch(category, **kwargs)
+
+        return items
+
+    def fetch_items(self, category, **kwargs):
+        """Fetch Change items
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:123]: `Change` should go in lower-case.

File: perceval/backends/core/weblate.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+        search_fields = {
+            DEFAULT_SEARCH_FIELD: self.metadata_id(item)
+        }
+
+        return search_fields
+
+    def fetch(self, category=CATEGORY_CHANGES, from_date=DEFAULT_DATETIME):
+        """Fetch changes from Weblate API.
+
+        The method retrieves the changes that occurred on a project via the Weblate API.
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:106]: I would reword this to `..the activity that occurred...`

File: perceval/backends/core/weblate.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+
+    def fetch_items(self, category, **kwargs):
+        """Fetch Change items
+
+        :param category: the category of items to fetch
+        :param kwargs: backend arguments
+
+        :returns: a generator of items
+        """
+        logger.info("Fetching data")
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:130]: Can we improve this message with something like:
`logger.info("Fetching %s items on '%s' from '%s", category, self.url, from_date)`. You will need to put this under the `from_date` variable.

File: perceval/backends/core/weblate.py [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+    def has_resuming(cls):
+        """Returns whether it supports to resume the fetch process.
+
+        :returns: this backend supports items resuming
+        """
+        return True
+
+    @staticmethod
+    def metadata_id(item):
+        """Extracts the identifier from a Change item."""
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:165]: This should only be `an item`.

File: perceval/backends/core/weblate.py [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+
+        return str(item['id'])
+
+    @staticmethod
+    def metadata_updated_on(item):
+        """Extracts the update time from a Change item.
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:171]: Say only `an item`.

File: perceval/backends/core/weblate.py [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+
+        The timestamp is based on the current time when the hit was extracted.
+        This field is not part of the data provided by Google API. It is added
+        by this backend.
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:175]: This message is completely wrong and references `googlehits`. Please make reference about how the date is obtained here.

File: perceval/backends/core/weblate.py [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+        ts = str_to_datetime(ts)
+
+        return ts.timestamp()
+
+    @staticmethod
+    def metadata_category(item):
+        """Extracts the category from a GoogleHits item.
+
+        This backend only generates one type of item which is
+        'google_hits'.
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:191]: This makes reference to googlehits.

File: perceval/backends/core/weblate.py [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+
+        if self.project:
+            path = urijoin(self.base_url, 'projects', self.project, 'changes')
+        else:
+            path = urijoin(self.base_url, 'changes')
+
+        return self.fetch_items(path, payload)
+
+    def fetch_items(self, path, payload):
+        """Return the items from github API using links pagination"""
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:283]: Please check this comment, it makes reference to GitHub.

File: perceval/backends/core/weblate.py [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+
+        parser = BackendCommandArgumentParser(cls.BACKEND,
+                                              archive=True,
+                                              from_date=True,
+                                              ssl_verify=True,
+                                              token_auth=True)
+
+        group = parser.parser.add_argument_group('Weblate arguments')
+        # Weblate options
+        group = parser.parser.add_argument_group('Weblate arguments')
```
Comment (sduenas) [inline perceval/backends/core/weblate.py:353]: Same line appears twice.

File: tests/data/weblate/weblate_changes.json [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+      "author": "https://my.weblate.org/api/users/2",
+      "timestamp": "2019-07-11T16:21:11.939033Z",
+      "action": 50,
+      "target": "",
+      "id": 2,
+      "action_name": "Created project",
+      "url": "https://my.weblate.org/api/changes/2/"
+    }
+  ]
+}
```
Comment (sduenas) [inline tests/data/weblate/weblate_changes.json:35]: Add an empty line at the end of the file.

File: tests/data/weblate/weblate_changes_archived.json [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+      "author": null,
+      "timestamp": "2020-07-11T17:21:04.609924Z",
+      "action": 21,
+      "target": "",
+      "id": 4,
+      "action_name": "Rebased repository",
+      "url": "https://my.weblate.org/api/changes/4/"
+    }
+  ]
+}
```
Comment (sduenas) [inline tests/data/weblate/weblate_changes_archived.json:62]: Add an empty line at the end of the file.

File: tests/data/weblate/weblate_changes_expected.json [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+            "author": null,
+            "timestamp": "2020-07-11T17:21:04.609924Z",
+            "action": 21,
+            "target": "",
+            "id": 4,
+            "action_name": "Rebased repository",
+            "url": "https://my.weblate.org/api/changes/4/"
+        }
+    }
+]
```
Comment (sduenas) [inline tests/data/weblate/weblate_changes_expected.json:196]: Add an empty line at the end of the file.

File: tests/data/weblate/weblate_changes_page_2.json [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+      "author": null,
+      "timestamp": "2020-07-11T17:21:04.609924Z",
+      "action": 21,
+      "target": "",
+      "id": 4,
+      "action_name": "Rebased repository",
+      "url": "https://my.weblate.org/api/changes/4/"
+    }
+  ]
+}
```
Comment (sduenas) [inline tests/data/weblate/weblate_changes_page_2.json:34]: Add an empty line at the end of the file.

File: tests/data/weblate/weblate_user_no_permission.json [hunk 17]
```diff
@@ -0,0 +1,3 @@
+{
+    "detail": "Not found."
+}
```
Comment (sduenas) [inline tests/data/weblate/weblate_user_no_permission.json:3]: Add an empty line at the end of the file.

**Review by sduenas — CHANGES_REQUESTED** [review]
In overall, the patch looks ok. Please fix the minor comments I added to this review.

File: perceval/backends/core/weblate.py (hunk 2 above)
Comment (zhquan) [reply perceval/backends/core/weblate.py:42]: The name of the category is `changes` why should we call it `change`?

**Comment by sduenas** [comment]
The repository https://github.com/chaoss/grimoirelab-perceval-weblate already includes this backend.
